File: crates/pyxis/src/angle.rs
/// Converts a `f64` angle value (in decimal degrees) to degrees, minutes, and seconds (DMS).
///
/// # Example
/// ```
/// use float_cmp::assert_approx_eq;
/// let angle: f64 = 30.76;
/// let dms = pyxis::angle_to_dms(angle);
/// assert_eq!(dms.0, 30);
/// assert_eq!(dms.1, 45);
/// assert_approx_eq!(f64, dms.2, 36.0, epsilon = 1e-6);
/// ```
pub fn angle_to_dms(angle: f64) -> (i32, i32, f64) {
    let degree = angle.trunc() as i32; // Get degrees
    let minutes_float = (angle - degree as f64) * 60.0; // Get fractional part and convert to minutes
    let minute = minutes_float.trunc() as i32; // Get minutes as integer
    let second = (minutes_float - minute as f64) * 60.0; // Get seconds
    (degree, minute, second)
}

/// Converts the DMS (degrees, minutes, seconds) format back into a decimal degree value.
///
/// # Returns
///
/// A `f64` representing the angle in decimal degrees.
///
/// # Example
/// ```
/// use float_cmp::assert_approx_eq;
/// let angle = pyxis::dms_to_angle(30,45,36.0);
/// assert_approx_eq!(f64, angle, 30.76, epsilon = 1e-6);
/// ```
pub fn dms_to_angle(degree: i32, minute: i32, second: f64) -> f64 {
    degree as f64 + minute as f64 / 60.0 + second / 3600.0
}

File: python/pyxis/_angle.py
from typing import Any

import numpy as np


def angle2dms(angle) -> tuple[Any, Any, Any]:
    """
    Convert float angle to degree, minute, second.

    Parameters
    ----------
    angle
        Float angle.

    Returns
    -------
    tuple[Any,Any,Any]
        Degree, minute, second.
    """
    deg = np.trunc(angle)
    min = np.trunc((angle - deg) * 60.0)
    sec = (angle - deg - min / 60.0) * 3600.0
    return deg, min, sec


def dms2angle(deg, min, sec):
    """
    Convert degree, minute, second to float angle.

    Parameters
    ----------
    deg, min, sec
        Degree, minute, second.

    Returns
    -------
    Any
        Float Angle.

    Examples
    --------
    >>> from glatzel import pyxis
    >>> pyxis.dms2angle(30, 45, 36)
    30.76
    """
    return deg + min / 60.0 + sec / 3600.0
